lib/scanner/xml-scanner/xml-scanner.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2018 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#ifndef XML_SCANNER_H_INCLUDED
+#define XML_SCANNER_H_INCLUDED
+
+#include "syslog-ng.h"
+#include "messages.h"
+#include "logmsg/logmsg.h"
+
+#include <string.h>
+
+typedef void (*PushCurrentKeyValue)(const gchar *name, const gchar *value, gssize value_length, gpointer user_data);
+
+typedef void (*StartElement) (GMarkupParseContext *context, const gchar *element_name, const gchar **attribute_names,
+                              const gchar **attribute_values, gpointer user_data, GError **error);
+typedef void (*EndElement) (GMarkupParseContext *context, const gchar *element_name, gpointer user_data,
+                            GError **error);
+typedef void (*Text) (GMarkupParseContext *context, const gchar *text, gsize text_len, gpointer user_data,
+                      GError **error);
+
+typedef struct
+{
+  gboolean strip_whitespaces;
+  GList *exclude_tags;
+} XMLScannerOptions;
+
+typedef struct
+{
+  GMarkupParseContext *xml_ctx;
+  GMarkupParser scanner_callbacks;
+  XMLScannerOptions *options;
+  gboolean pop_next_time;
+  gboolean matchstring_shouldreverse;
+  GPtrArray *exclude_patterns;
+  PushCurrentKeyValue push_function;
+  StartElement start_element_function;
+  EndElement  end_element_function;
+  Text  text_function;
+  gpointer user_data;
+  GString *key;
+  GString *text;
+} XMLScanner;
```

Comment:
What do you think about this below?

```
typedef void (*PushCurrentKeyValueCB)(const gchar *name, const gchar *value, gssize value_length, gpointer user_data);

typedef struct
{
  PushCurrentKeyValueCB func;
  gpointer user_data;
} PushCurrentKeyValue;

typedef struct
{
  XMLScannerOptions *options;
  gboolean pop_next_time;
  gboolean matchstring_shouldreverse;
  GString *key;
  GString *text;

  StartElement start_element_function;
  EndElement end_element_function;
  Text  text_function;

  GPtrArray *exclude_patterns;
  PushCurrentKeyValue push_key_value;
} XMLScanner;
```

Reply:
Okay, we talked in person that the class can be a bit huge (not anymore since g_markup is removed from the class!! ), but .e.g `user_data` can be confusing where is it used.
A bit grouping can help, so thanks for the tip!